runtime/skiplist.c

```diff
@@ -0,0 +1,214 @@
+/*                                                                        */
+/*   Copyright 2020 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+/* A dictionary data structure implemented as skip lists
+   (see William Pugh, "Skip lists: a probabilistic alternative to
+   balanced binary trees", Comm. ACM 33(6), 1990). */
+
+#include <stddef.h>
+#include "caml/config.h"
+#include "caml/memory.h"
+#include "caml/misc.h"
+#include "caml/skiplist.h"
+
+/* Generate a random level for a new node: 0 with probability 3/4,
+   1 with probability 3/16, 2 with probability 3/64, etc.
+   We use a simple linear congruential PRNG (see Knuth vol 2) instead
+   of random(), because we need exactly 32 bits of pseudo-random data
+   (i.e. 2 * (NUM_LEVELS - 1)).  Moreover, the congruential PRNG
+   is faster and guaranteed to be deterministic (to reproduce bugs). */
+
+static uint32_t random_seed = 0;
+
+static int random_level(void)
+{
+  uint32_t r;
+  int level = 0;
+
+  /* Linear congruence with modulus = 2^32, multiplier = 69069
+     (Knuth vol 2 p. 106, line 15 of table 1), additive = 25173. */
+  r = random_seed = random_seed * 69069 + 25173;
+  /* Knuth (vol 2 p. 13) shows that the least significant bits are
+     "less random" than the most significant bits with a modulus of 2^m,
+     so consume most significant bits first */
+  while ((r & 0xC0000000U) == 0xC0000000U) { level++; r = r << 2; }
+  CAMLassert(level < NUM_LEVELS);
+  return level;
+}
+
+/* Initialize a skip list */
+
+void caml_skiplist_init(struct skiplist * sk)
+{
+  sk->forward[0] = NULL;
+  sk->level = 0;
+}
+
+/* Search a skip list */
+
+int caml_skiplist_find(struct skiplist * sk, uintnat key, uintnat * data)
+{
+  int i;
+  struct skipcell * e, * f;
+
+  e = (struct skipcell *) sk;
+  for (i = sk->level; i >= 0; i--) {
+    while (1) {
+      f = e->forward[i];
+      if (f == NULL || f->key >= key) break;
+      e = f;
+    }
+  }
+  e = e->forward[0];
+  if (e != NULL && e->key == key) {
+    *data = e->data;
+    return 1;
+  } else {
+    return 0;
+  }
+}
+
+int caml_skiplist_find_below(struct skiplist * sk, uintnat k,
+                             uintnat * key, uintnat * data)
+{
+  int i;
+  struct skipcell * e, * f;
+
+  e = (struct skipcell *) sk;
+  for (i = sk->level; i >= 0; i--) {
+    while (1) {
+      f = e->forward[i];
+      if (f == NULL || f->key >= k + 1) break;
```

Comment:
Why is this `>= k + 1`, rather than `> k` (I thought I saw a reason, then doubt it, so just asking instead!)

Reply:
Historical reasons, sort of.  At some point I thought "it would be simpler to search for the key immediately after the one we're interested in", that is, `k  + 1` instead of `k`.  But of course `> k` is better than `>= k + 1` if only because it avoids overflow.